Support the platform specific authentication of krane in "auth get" command

ingwonsong opened the pull request:
This PR enables to work "auth get" command with the platform specific keychains given in `krane` tool.

imjasonh commented:
This is really interesting! So the idea is that with this change, `krane auth get gcr.io` will print out the auth token, possibly as derived from a compiled-in cred helper like ecr-login etc.

I think we'll want a simple end-to-end test of this behavior, for [GHCR](https://github.com/google/go-containerregistry/blob/main/.github/workflows/ghcr-auth.yaml) and [ECR](https://github.com/google/go-containerregistry/blob/main/.github/workflows/ecr-auth.yaml) at least. 

For ECR, that means calling `krane auth get ${{ env.AWS_ACCOUNT }}.dkr.ecr.${{ env.AWS_REGION }}.amazonaws.com` and checking that it doesn't fail, and that it prints something -- though we should be careful not to print that out to logs.

Could you add that to this PR?

ingwonsong commented:
@imjasonh Sure. I have just added the tests. 
In addition, I also added small extension.

Previously, `crane` and `krane` get a registry address through `stdio` only, like:
```
$ echo "example.registry.com" | krane auth get
```

But, I added to another way to specify the registry address as an argument as follows:
```
$ krane auth get example.registry.com
```

jonjohnsonjr commented on cmd/crane/cmd/auth.go:
```diff
@@ -62,30 +63,46 @@ func toCreds(config *authn.AuthConfig) credentials {
 				return err
 			}
-			authorizer, err := authn.DefaultKeychain.Resolve(reg)
+			keychain := authn.DefaultKeychain
+			if options != nil {
```
I think you can drop this and just always pull out `crane.GetOptions(*options...).Keychain` given that crane options use authn.DefaultKeychain as the... default keychain.

I'd just make sure nothing is passing nil options into here.

ingwonsong replied:
Good idea. Done.

jonjohnsonjr commented on cmd/gcrane/main.go:
```diff
@@ -42,7 +42,7 @@ func main() {
 
 	// Add or override commands.
-	gcraneCmds := []*cobra.Command{gcmd.NewCmdList(), gcmd.NewCmdGc(), gcmd.NewCmdCopy(), cmd.NewCmdAuth("gcrane", "auth")}
+	gcraneCmds := []*cobra.Command{gcmd.NewCmdList(), gcmd.NewCmdGc(), gcmd.NewCmdCopy(), cmd.NewCmdAuth(nil, "gcrane", "auth")}
```
I'd pull the options from line 42 out and reuse them here so that `gcrane auth get` uses `google.Keychain`.

ingwonsong replied:
Done.

Now, `gcrane auth get` also works!

jonjohnsonjr commented:
Thanks for this change, this is awesome.

If I had the time, I would set up workload identity so that we could test `gcrane auth get` and `krane auth get` against GCP, but I'm fine without it 😅 

Looks like you just need to regenerate the docs for the presubmit.
